src/Grpc/JsonTranscoding/src/Microsoft.AspNetCore.Grpc.JsonTranscoding/Internal/Json/MessageTypeInfoResolver.cs

```diff
@@ -0,0 +1,165 @@
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections;
+using System.Diagnostics.CodeAnalysis;
+using System.Text.Json;
+using System.Text.Json.Serialization.Metadata;
+using Google.Protobuf;
+using Google.Protobuf.Reflection;
+using Grpc.Shared;
+using Type = System.Type;
+
+namespace Microsoft.AspNetCore.Grpc.JsonTranscoding.Internal.Json;
+
+internal sealed class MessageTypeInfoResolver : IJsonTypeInfoResolver
+{
+    private readonly JsonContext _context;
+
+    public MessageTypeInfoResolver(JsonContext context)
+    {
+        _context = context;
+    }
+
+    public JsonTypeInfo? GetTypeInfo(Type type, JsonSerializerOptions options)
+    {
+        if (!IsStandardMessage(type, out var messageDescriptor))
+        {
+            return null;
+        }
+
+        var typeInfo = JsonTypeInfo.CreateJsonTypeInfo(type, options);
+        typeInfo.CreateObject = () => Activator.CreateInstance(type)!;
+
+        var fields = messageDescriptor.Fields.InFieldNumberOrder();
+        var mappings = CreateJsonFieldMap(fields);
+
+        foreach (var field in fields)
+        {
+            mappings.Remove(field.JsonName);
+
+            var propertyInfo = CreatePropertyInfo(typeInfo, field.JsonName, field, isWritable: true);
+            typeInfo.Properties.Add(propertyInfo);
+        }
+
+        // Fields have two mappings: the original field name and the camelcased JSON name.
+        // The JSON name can also be customized in proto with json_name option.
+        // Add extra setter only properties for mappings that haven't already been added.
+        foreach (var mapping in mappings)
+        {
+            var propertyInfo = CreatePropertyInfo(typeInfo, mapping.Key, mapping.Value, isWritable: false);
+            typeInfo.Properties.Add(propertyInfo);
+        }
+
+        return typeInfo;
+    }
+
+    private static bool IsStandardMessage(Type type, [NotNullWhen(true)] out MessageDescriptor? messageDescriptor)
+    {
+        if (!typeof(IMessage).IsAssignableFrom(type))
+        {
+            messageDescriptor = null;
+            return false;
+        }
+
+        messageDescriptor = JsonConverterHelper.GetMessageDescriptor(type);
+        if (messageDescriptor == null)
+        {
+            return false;
+        }
+
+        // Wrappers and well known types are handled by converters.
+        if (ServiceDescriptorHelpers.IsWrapperType(messageDescriptor))
+        {
+            return false;
+        }
+        if (JsonConverterHelper.WellKnownTypeNames.ContainsKey(messageDescriptor.FullName))
+        {
+            return false;
+        }
+
+        return true;
+    }
+
+    private JsonPropertyInfo CreatePropertyInfo(JsonTypeInfo typeInfo, string name, FieldDescriptor field, bool isWritable)
+    {
+        var propertyInfo = typeInfo.CreateJsonPropertyInfo(
+            JsonConverterHelper.GetFieldType(field),
+            name);
+
+        if (isWritable)
```

Comment:
nit: I was expecting the same as @krwq. Maybe change the parameter name to something like `isSerializable`